NAS-137494 / None / Update to 6.18 LTS

ixhamza opened the pull request:
### Description

Rebased all TrueNAS patches from truenas/linux-6.12 onto Linux v6.18 release tag. Discarded merge commits, changelog updates, and kernel version bump commits. Several patches were already upstreamed and excluded from this rebase.

**Patches with merge conflicts resolved:**
| Commit | Description | Conflicts |
|--------|-------------|-----------|
| 3ac7142 | efi: Add an EFI_SECURE_BOOT flag to indicate secure boot mode | *include/linux/efi.h* |
| 57ac602 | block: Introduce queue limits and sysfs for copy-offload support | *block/blk-settings.c, block/blk-sysfs.c* |
| 35c569e5 | disable multiple device (md) driver | *scripts/package/truenas/truenas.config* |
| eba7974 | NVME Encapsulation support for Trimode HBA | *drivers/scsi/mpt3sas/mpt3sas_ctl.h* |
| 7d500ea | fs/cifs - add ZFS ACL support to SMB client | *fs/smb/client/cifs_debug.c* |
| 6d72bc4 | Add DACL support to nfsd (v4.1+) | *fs/nfsd/nfs4xdr.c* |
| 8a3c5b3 | nvme: skip optional id ctrl csi for versions less than 2.0.0 | *drivers/nvme/host/core.c* |
| ea6a273 | ixgbe: Print fw warning message only once | *drivers/net/ethernet/intel/ixgbe/* |
| 49dc703 | Build production and debug kernels for SCALE | *scripts/package/mkdebian* |
| 998ce51 | truenas-scale: Add NTB to default tn.config | *arch/x86/Kconfig* |
| e1969f6 | Add initial support for large xattrs | *fs/xattr.c* |
| 46e77e3 | Add TrueNAS Debian build customizations | *Makefile, scripts/package/* |

**New commits for 6.18 compatibility:**
- **cbca2395ce4ce096d8bd7af15d4cd9c6cdabfda4**: Imported Debian v6.18-rc7 configuration (will update to v6.18 release config once published by Debian)
- CI improvements:
  - **3708feb456f63fb1035eeaecd3f5cac098125b6f**: Added aggressive disk cleanup for larger 6.18 build artifacts
  - **afd59a7b52a0a43e5f3c05da619a9d2189b6d643**: Updated build dependencies (`libdw-dev`)
  - **c39b563a9d2f96b652c894a16dc42676cd4635a2**: Enabled workflow cancellation for outdated runs
- **11b74375e1e58d1c60aaee579e229ac45e22155c**: Standardized `CONFIG_TRUENAS` macro usage to `#ifdef` throughout codebase
- **eb7dc66a66deb5306200f28180fd92f7c2d110c7**: Enabled `CONFIG_NFSD_LEGACY_CLIENT_TRACKING` for API test compatibility (Debian disabled in 6.18)
- **8d5020feec2a7aceec2e302f05f1d4612ab08d1b**: Disabled `CONFIG_TRACEFS_AUTOMOUNT_DEPRECATED` to suppress kernel warning spam
- **6a0b8768408aae9edb7ef32fabd948857d8258a4**: Fix Samba Client regression with kernel 6.18+

**Related PRs**
- **[scale-build](https://github.com/truenas/scale-build/pull/956):**
  - Switch kernel branch from truenas/linux-6.12 to truenas/linux-6.18
  - Upgrade NVIDIA driver from 580 to 590 for 6.18 compatibility
  - Add zstd package for kernel compression
- **[middleware](https://github.com/truenas/middleware/pull/17780):**
  - Fix mount tree parsing for kernel 6.18 mount ID allocation changes
- **[ZFS](https://github.com/truenas/zfs/pull/320):**
  - Update META to enable 6.18 kernel compatibility


### Testing
- [Scale Build](http://jenkins.eng.ixsystems.net:8080/job/master/job/custom/1739/)
- [API Tests](http://jenkins.eng.ixsystems.net:8080/job/tests/job/api_tests/6875/)
- CI Testing: Production and debug kernel builds complete successfully
- Boot testing: Confirmed successful boot on KVM with no kernel regressions
- NVIDIA driver 590.44.01 compiles successfully on 6.18
- Middleware mount tree parsing tested on both 6.12 and 6.18
- API test `test_300_nfs test_state_directory` failure resolved by enabling `CONFIG_NFSD_LEGACY_CLIENT_TRACKING`

dhowells committed:
efi: Add an EFI_SECURE_BOOT flag to indicate secure boot mode

UEFI machines can be booted in Secure Boot mode.  Add an EFI_SECURE_BOOT
flag that can be passed to efi_enabled() to find out whether secure boot is
enabled.

Move the switch-statement in x86's setup_arch() that inteprets the
secure_boot boot parameter to generic code and set the bit there.

Suggested-by: Ard Biesheuvel <ard.biesheuvel@linaro.org>
Signed-off-by: David Howells <dhowells@redhat.com>
Reviewed-by: Ard Biesheuvel <ard.biesheuvel@linaro.org>
cc: linux-efi@vger.kernel.org
[rperier: Forward-ported to 5.5:
 - Use pr_warn()
 - Adjust context]
[bwh: Forward-ported to 5.6: adjust context]
[bwh: Forward-ported to 5.7:
 - Use the next available bit in efi.flags
 - Adjust context]

bwhacks committed:
efi: Lock down the kernel if booted in secure boot mode

Based on an earlier patch by David Howells, who wrote the following
description:

> UEFI Secure Boot provides a mechanism for ensuring that the firmware will
> only load signed bootloaders and kernels.  Certain use cases may also
> require that all kernel modules also be signed.  Add a configuration option
> that to lock down the kernel - which includes requiring validly signed
> modules - if the kernel is secure-booted.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
[Salvatore Bonaccorso: After fixing https://bugs.debian.org/956197 the
help text for LOCK_DOWN_IN_EFI_SECURE_BOOT was adjusted to mention that
lockdown is triggered in integrity mode (https://bugs.debian.org/1025417)]
Signed-off-by: Salvatore Bonaccorso <carnil@debian.org>

anodos325 committed:
Add initial support for large xattrs (#1)

Support for alternate datastreams over the SMB protocol has been
historically enabled in  such a way that Samba writes them as
filesystem extended attributes in the user namespace. FreeBSD has no
practical limit on xattr size, and so clients (often MacOS) may write
ones that exceed the 64 KiB limit imposed by the Linux kernel. Since
XATTR_SIZE_MAX is uesd in many places in the kernel, and not all
filesystems support large xattrs, introduce new constant
XATTR_LARGE_SIZE_MAX that is used as an alternate value if the
filesystem sb_flags has SB_LARGEXATTR. There will be corresponding
commit in ZFS to set this flag when it is defined and xattrs are
enabled on the ZFS dataset.

This commit also introduces flag SB_NFSV4ACL which will be used
to indicate and enable NFSv4-specific behavior in kernel with regard
to permissions.

These new features / alternate behavior are controlled by the
compile-time kernel compilation flag CONFIG_TRUENAS, which defaults
to n (off). In principle, TrueNAS-specific changes that deviate from
a vanilla Linux kernel can be removed for testing purposes by changing
CONFIG_TRUENAS=n in the relevant build scripts.

Signed-off-by: Andrew Walker <awalker@ixsystems.com>

Commit:
ntb/plx: Add support for PLX hardware

Added initial support for PLX Non-Transparent Bridge.

Commit:
truenas-scale: Add NTB to default tn.config

Enable NTB and NTB tools in the Truenas config.  In addition, enable
the Intel NTB driver, so that we have at least one NTB driver available.

anodos325 committed:
Add nfsv4-specific permissions checks to kernel (#2)

There are various places in which evaluation of permissions
in the presence of an NFSv4 ACL is more nuanced than what is
typical when evaluating traditional POSIX permissions. For
example, a user may be permitted to delete a file if he
has DELETE permissions on the file or DELETE_CHILD permissions
on the parent directory. Traditional POSIX permissions will
only check for MAY_WRITE | MAY_EXEC on parent directory.

Several new inode permissions masks have been added to facilitate these
NFSv4-specific checks corresponding to different NFSv4 permissions
that grant abilities to make changes to files. For the purpose of
this commit and the goal of providing rough a approximation of
NFSv4 access checks, only write (and not read) access checks have
been implemented. This is selectively done in a way to grant
minimal compliance with permissions as defined in RFC-5661.

The new permissions-related behavior is only applied when the
inode sb_flag SB_NFS4ACL is present. In this case, the onus of full
implementation of requisite features to satisfy the ACL behavior
specified in RFC-5661 is delegated to the filesystem's inode
permissions interface (i_op->permission). If possible we try to
check for the convention POSIX permission first before trying
the NFSv4-equivalent. For example, when writing an xattr, we
check for WRITE_DATA before WRITE_NAMED_ATTRS because in the case
of former with a trivial ACL we can avoid having to evaluate the
full ACL, and instead merely look at POSIX mode.

Commit:
Remove Chelsio FCoE options from kernel config

csiostor seems to cause Chelsio T6 firmware to crash.

Jira: NAS-110910

amotin committed:
Add /sys/kernel/wait_for_device_probe attribute.

Being written anything waits for all device probe to complete before
returning.  After that `udevadm settle` used by ZFS scripts really
can provide system with all disks detected for boot pool import.

Ticket:	NAS-108200

amotin committed:
NTB: ntb_transport: Allow early client registration

Before this change it was impossible to load client modules before
NTB hardware is probed.  This change removes the limitation.  New
NTB transports will get their children devices as they come in.

amotin committed:
ntb_transport: Port some fixes from FreeBSD.

amotin committed:
ntb: Provide more reasonable interrupt names.

amotin committed:
ntb: Mask doorbells while processing them.

This fixes interrupt storms on hardware using legacy level-triggered
interrupts, since doorbell processing could take time after interrupt
handler completion, that triggered extra interrupts in a loop.

amotin committed:
ntb_split: Introduce new driver to split NTB resources.

It attaches as client to some NTB device and provides several other NTB
device, splitting NTB resources according to config module parameter.
For example, one BAR can be dedicated for remote memory access, while other
resources can be used for packet transport for virtual Ethernet interface.

Commit:
Setup CI for TrueNAS/Linux repository

If a previous successful run is
present, then skip re-run for its pull request.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

Commit:
Better Linux kernel configuration procedure

Previously for TrueNAS, Debian Linux kernel configuration was used and
TrueNAS config options were added on top of that. Because of that,
TrueNAS kernel config in 'scripts/package/truenas/tn.config' has grown
very large and difficult to manage for TrueNAS only options.

Debian Linux kernel configuration for version 6.1.55 has been added
as 'debian_amd64.config' to keep the options from Debian seperate from
TrueNAS options. TrueNAS only config options are stored in
'truenas.config'.

Kernel configuration can now be generated as:

	1) make ARCH=x86_64 defconfig
	2) ./scripts/kconfig/merge_config.sh .config \
			./scripts/package/truenas/debian_amd64.config
	3) ./scripts/kconfig/merge_config.sh .config \
			./scripts/package/truenas/truenas.config
	4) ./scripts/kconfig/merge_config.sh .config \
			./scripts/package/truenas/debug.config

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

amotin committed:
Introduce driver_override for NTB devices.

Without this it was impossible to use multple NTB consumer drivers,
since kernel just attached the first one to all NTBs.  This replicates
driver_override device attribute from PCI, plus adds module parameter
to ease default setting.

Signed-off-by: Alexander Motin <mav@FreeBSD.org>

amotin committed:
Add compact scraptchpad protocol for ntb_transport.

Previously ntb_transport required at least 6 scratchpad registers, plus
2 more for each additional memory window.  That is too much for some
configurations, where several drivers have to share resources of the same
NTB hardware.  This patch introduces new compact version of the protocol,
requiring only 3 scratchpad registers, plus one more for each additional
memory window.  The optimization is based on fact that neither of version,
number of windows or number of queue pairs really need more then one byte
each, and window sizes of 4GB are not very useful now.  The new protocol
is activated automatically when the configuration is low on scratchpad
registers, or it can be activated explicitly with module parameter.

Commit:
ahciem: Emulate SES enclosure for AHCI enclosures

Mostly complete.  AHCI controllers supporting enclosure management will
create a virtual SES logical unit on a dedicated SCSI host, allowing
control/status of EM LEDs.  link to the device in a given slot is created for
mapping the slot to a device:

```
root@trunasmini:~# strings /sys/class/enclosure/*/Slot*/device/inquiry
ATA     Samsung SSD 860 1B6Q
ATA     Samsung SSD 840 6B0Q
ATA     SanDisk SD6SB1M0600
ATA     SanDisk SDSSDHII00RL
root@trunasmini:~# ls /sys/class/enclosure/*/Slot*/device/block/
'/sys/class/enclosure/4:0:0:0/Slot 00/device/block/':
sda

'/sys/class/enclosure/4:0:0:0/Slot 01/device/block/':
sdb

'/sys/class/enclosure/4:0:0:0/Slot 02/device/block/':
sdc

'/sys/class/enclosure/4:0:0:0/Slot 03/device/block/':
sdd
root@trunasmini:~# ls /sys/block/*/device/enclosure_device:*
'/sys/block/sda/device/enclosure_device:Slot 00':
active  device  fault  locate  power  power_status  slot  status  type  uevent

'/sys/block/sdb/device/enclosure_device:Slot 01':
active  device  fault  locate  power  power_status  slot  status  type  uevent

'/sys/block/sdc/device/enclosure_device:Slot 02':
active  device  fault  locate  power  power_status  slot  status  type  uevent

'/sys/block/sdd/device/enclosure_device:Slot 03':
active  device  fault  locate  power  power_status  slot  status  type  uevent
```

Tasks left for future work:
 * avoid the layering violation in the SES driver
 * provide an option for the tunable to force SES emulation without LED support
 * make use of ALLOCATION LENGTH

Commit:
Allow concurrent skipping for CI

Currently, the CI run for Pull Requests are skipped only if a
successful Push CI run is found. This implies that Push CI run has to
complete before Pull Request should be opened, otherwise Pull Request
CI will also start in parallel to Push CI run.

If upon opening a Pull Request, a Push CI run is already in progress,
concurrent skipping allows to skip Pull Request CI run completely.

'same_content_newer' ensures that atleast one of Push or Pull Request
is run. If a new commit is added after opening a Pull Request or in
case of force push, atleast one of Pull Request or Push CI run will
be performed.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

Commit:
Build production and debug kernels for SCALE

This commit adds the support for production and debug kernels for SCALE.
Following is a summary of changes:

  - tn-production.config has been added. This fragment will be used to
    disable unnecessary debug option that are not required in production.
  - tn-debug.config has been added. This fragment will enable extra
    debug options to aid debugging during boot and runtime.
  - Unnecessary hardware options like WiFi, Bluetooth, NFSC etc. have
    been disabled for both production and debug kernels.
  - The packaging bits in kernel have been updated to include the extra
    version string in package name so that the packages for debug and
    production kernels can be uniquely identified.
  - A new changelog entry will be generated based on the updated package
    name for production and debug kernels.
  - EXTRAVERSION variable con be configured from the environment.
    `export EXTRAVERSION=-production` would build the packages for
    production kernel. This should be the first step to build the kernel
    packages moving forward.
  - GitHub Actions are updated to build both production and debug
    kernels.

A minor update has been made here while porting from 6.1, EXTRAVERSION
variable is set to '-production'. So by default, production kernel will
be built and we don't need to export the EXTRAVERSION variable while
building the production kernel.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

amotin committed:
Introduce NVDIMM NTB mirroring driver.

It is a fully compatible port of respective functionality from FreeBSD
TrueNAS Core.  On Linux it handles only NVDIMMs without labels, i.e.
only "raw" namespaces, covering whole range.  Instead the driver writes
own label into the last page of NVDIMM, using it to store disk serial
number and some information required for data synchronization.

amotin committed:
Add AMD NTB driver to the build as a module.

Commit:
Change SATA mobile LPM policy to max_performance

Debian set this to med_performance_with_dipm, which causes issues with
some hardware.  Set it to max_performance instead.

Signed-off-by: Ryan Moeller <ryan@iXsystems.com>

amotin committed:
Add DMA support to NVDIMM pmem driver.

Use dmaengine KPI to program I/OAT (or any other engine) to offload
memory copies for I/Os above certain size (smaller I/Os are faster
in software).  For systems configured for HA synchronization via
NTB use second DMA channel to duplicate all writes to remote NVDIMM.

While there, remove redundant loops from software read/write routines.
bio_for_each_segment() unlike bio_for_each_bvec() by itself splits
all the bio vectors into separate pages.  Assert it just in case.

amotin committed:
Workaround NVDIMM bugs of Supermicro X11DPI-NT boards.

First, Linux expects NVDIMM _DSM method with unknown UUID to return
zero, meaning no functions supported.  Supermicro though returns 1,
meaning error has happened.  It makes Linux think that NVDIMM supports
all command sets, selecting Intel as first on the list.  Workaround
it by checking actual function bits instead of bit 1, same as we are
doing on FreeBSD.

Second, Linux checks for _LSI/_LSR/_LSW methods presence to detect
NVDIMM's support for labels.  Supermicro though has those methods,
but they just always return error status.  It makes Linux disable
the NVDIMM after failed attempt to read its label during the probe.
Workaround it by actually calling _LSI method once and checking the
returned status.

Together those fix `ndctl list -D --health` for NVDIMMs implementing
Microsoft command set on the specified motherboards.

amotin committed:
NTB: ntb_transport: Remove children before destruction.

Children drivers must call their ntb_transport_free_queue() before
we clean up what's left (nothing should).  Otherwise we get double
free.

amotin committed:
NTB: ntb_transport: Remove node debugfs on destruction.

Otherwise on next attach we get warning message that the directory
already exist.

Commit:
Disable default enablement of init_on_alloc feature

Disable init_on_alloc to improve ZFS performance. ZFS allocates
and frees pages frequently, and init_on_alloc zeroes out the pages
during allocation. Disabling init_on_alloc should improve ZFS
performance.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

amotin committed:
pciehp: Ignore PCIe link status LT bit if DLLLA is set.

AMD NTB hot-plug sometimes reports Link Training while also reporting
Data Link Layer Link Active, which should be mutually exclusive.
Without this hot-plug fails sometimes.

bmeagherix committed:
NAS-119337 / 23.10 / Add ALUA support for iSCSI in SCALE (#67)

* scsi: core: sd: Add genhd_hidden flag

This flag will permit SCSI low level drivers to set the gendisk GENHD_FL_HIDDEN
flag, resulting in a gendisk that can be used for I/O submission inside the
kernel, but which is not registered as a user facing block device.

* scsi: iscsi_tcp: Hide logged-in HA peer node iSCSI targets from local access

ixhamza committed:
Add support to periodically poll enclosures

SES device of the enclosure reports disk's SAS address in Additional
Element Status few seconds after disk is detected by SES driver. The
periodic polling of enclosures allows enclosure to be added in case of
the lost events.

Commit:
Disable default restriction to access IO memory from /dev/mem

Allow userspace (root) access to all IO memory. /dev/mem allows
userspace access to memory rabges that may or may not be actively
by a driver. Required for plx_eeprom utility.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

ixhamza committed:
Add support for SATA disk resize without power cycle

Unlike SAS disks, the Linux kernel does not update the size of SATA
disks immediately unless the power cycle is done or the disk is
revalidated (due to some error condition). In this patch, we refresh
the sector count by revalidating the disk in response to ATA sector
update command.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

Commit:
Enable CONFIG_SUNRPC_DEBUG and CONFIG_NFS_DEBUG for production kernel

This commit enables CONFIG_SUNRPC_DEBUG and CONFIG_NFS_DEBUG for
production kernel. These configs are needed to enable the debugging
levels for NFS service.

These configurations are enabled for debug kernel.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

amotin committed:
ntb: intel: Set proper DMA masks, matching XLAT sizes.

Without this with BAR split activated ntb_transport allocates memory
that can't be handled by the 32bit XLAT registers.

Signed-off-by: Alexander Motin <mav@ixsystems.com>

ixhamza committed:
ixgbe: Print fw warning message only once

If FW_VAL_BIT is set to 0, a warning message is logged every second
in an ixgbe_service_task(), making the TrueNAS console unusable.
Some users still face the issue after updating to the latest Intel
firmware without noticing any performance issue in NIC. In the case
of this scenario, we should only log the warning message once. This
should still bring attention to the user and does not spam the logs.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

amotin committed:
Rename FORCE_MAX_ZONEORDER to ARCH_FORCE_MAX_ORDER.

This follows upstream 0192445cb2f7ed1cd7a95a0fc8c7645480baba25.

anodos325 committed:
Implement native NFSv4 ACLs in NFS server

ZFS currently exposes native ZFS nfsv4 ACL type through
the system.nfs4_acl_xdr xattr. underlying ACL type can
be determined via inode sb flags (previous kernel work).

This PR makes the Kernel NFS server aware of this ACL
type and has it read from and write to the xattr is
the relevant NFS40 ops.

At this point only NFS40 ACLs are implemented (RFC3530)
because they are currently only ones supported by the
Kernel NFS client in Linux.

ACE_INHERITED_ACE is returned in request for NFS40 ACL.
This is for consistency with FreeBSD behavior, and is
important for SMB / NFS compatibility. At a future
point in time, full NFS41 ACL compatibility will
most likely be required (for feature parity with
other commercial vendors). nfs4-acl-tools version 0.3.3
and earlier contains a bug where the persence of
ACE_INHERITED_ACE is mistakenly identified as making
ACE apply to the file owner.

Signed-off-by: Andrew Walker <awalker@ixsystems.com>

Commit:
Build perf package for TrueNAS kernel

Perf package is highly coupled with Linux kernel version present on
the system. The sources are present under tools/perf directory. The
packages available in upstream Debian repos are specific to Debian
Linux kernels and are not compatible with TrueNAS kernels due to
version mismatch.

For Bookworm, the stable kernel is v6.1 but it is hard to kep track
of kernel packages and update them when Debian does.

It seems we would have to build the perf package. There has been an
effort to upstream a patch that updates builddeb to generate a tools
package but it looks like it was not accepted for some reason and so
far it's being used to generate a perf/tools package with custom
kernels.

https://lists.debian.org/debian-kernel/2015/04/msg00013.html

Some adaptations have been made to this patch to make it compatible
with latest sources like installing the perf build dependencies to
the GitHub CI and updating the control file from mkdebian where
other parts of control file are being generated. Package name has
been updated to linux-perf-truenas so that it does not conflict
with upstream package in Debian repo.

Debian builds the perf package using it's own packaging bits, which
does not follow kernel provided packaging. Porting the debian
packaging bits here to build the perf package is not worth the
effort in my opinion.

Since both production and debug kernel are built from same sources
and perf binaries that are built, should be the same for both kernels,
perf package is only built with production kernel.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

Commit:
Disable unnecessary drivers for 6.1

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

ixhamza committed:
nvme: skip optional id ctrl csi for versions less than 2.0.0

The NVME_ID_CNS_CS_CTRL command has been introduced in version 2.0.0.
However, this command returns an "Invalid Field" error when executed
on previous NVMe versions. In the case of CM6 devices, this error is
logged on the error log page. Additionally, for CM6 drives, the
reservation commands manipulated by the fenced process indirectly
trigger the NVME_ID_CNS_CS_CTRL command every few seconds, leading
to a periodic increment in the error count.

While these error entries are harmless, smartctl periodically increases
the NVMe error counts. Smartctl displays the increase in NVMe error
count due to these errors on both the console and the log file,
without providing additional information. To resolve this issue,
a simple solution is to skip the NVME_ID_CNS_CS_CTRL command if the
version is less than 2.0.0.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

Commit:
Enable Media USB Support

A user reported that they are not able to use USB TV Tuner
device. Enabling Media USB support enables the drivers for
it.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

amotin committed:
ntb_hw_amd: Simulate link flap after unclean reboot

In case physical link events reporting on primary side is unreliable
it may happen that secondary side tries to renegotiate after reboot
while the primary side still thinks the link was up all the time and
nothing has happened.  To workaround that, on the driver load check
the READY bit in SIDEINFO register, and if it is already set, it
means we had a dirty shutdown.  In such case clear it and send the
doorbell interrupt to the other side, similar to driver unload.
After that wait for a second for remote side to see the change and
continue the initialization sequence.  If the physical link events
reporting is working, then remote side will clear the READY bit by
itself on link down event and this change will be a NOP.

Signed-off-by: Alexander Motin <mav@ixsystems.com>

amotin committed:
ntb_netdev: Allocate new skb on rx before using the old

This allows to call ntb_transport_rx_enqueue() with old skb in case
of allocation failure.  Otherwise receive dies after NTB_RXQ_SIZE
allocation failures due to lack of queued buffers.

Signed-off-by: Alexander Motin <mav@ixsystems.com>

bmeagherix committed:
Add flag to struct nfs4_acl to allow dacl support

Also move enum nfs4_acl_type for reuse.

bmeagherix committed:
Add DACL support to nfsd (v4.1+)

A DACL differs in size from an ACL by the addition of an acl_flag.

DACL support is modeled on the existing ACL support.  Some functions
(e.g. nfsd4_get_nfs4_acl) have their signatures widened to include a
acl_type parameter.

bmeagherix committed:
Move some nfs4_acl_xdr handling functions from nfsd to nfs_common for reuse

Also now use ACE4SIZE and add a check wrt remaining in
convert_nfs41xdr_to_nfs40_acl / convert_to_nfs40_ace

bmeagherix committed:
Add isdir parameter to generate_nfs41acl_buf and populate acl flags

bmeagherix committed:
Add nfs client-side support for NA41_NAME ("system.nfs4_acl_xdr")

The nfs4_xattr_list_nfs4_acl_xdr, nfs4_xattr_get_nfs4_acl_xdr and
nfs4_xattr_set_nfs4_acl_xdr functions are implemented using the existing
client-side support for DACL.

Currently only OWNER@, GROUP@, EVERYONE@ and numeric uids or gids are
supported in the ACEs.

Commit:
Remove tn.config and add CI fixes

This commit removes tn.config that was removed in v6.1 in commit
c17939b5ca583dfbd154ed7e6fd0d3a725307c32.

This commit provides addtional free space for CI runners from
commit c3e03a5d87caa6b89176e677ee5a7ce1aa2cb413.

bd0cd59c87e9469cd37f61c1b62ef3739296e183 is also applied here
to fix couple issues in CI.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

anodos325 committed:
NAS-122619 / 24.04/ fs/cifs - add ZFS ACL support to SMB client (#144)

Add an xattr handler with same namespace as ZFS ACL xattr handler
to allow userspace utilities to easily preserve and convert
contents of SMB Security Descriptor DACL into native ZFS ACL
when ingesting data during migration via SMB client (for example
using rsync with the explicit option to preserve the xattr in
question).

This PR also adds a new procfs endpoint:
/proc/fs/cifs/zfsacl_configuration_options that can be used to
control error handling for cases where we can't convert SID
into a Unix ID, and currently also whether we allow setting
NT ACL via xattr writes on remote SMB server (disabled by default).

Commit:
Disable DCB for Chelsio T4

DCB seems to be generating spam on the console for users with
Chelsio T4. Disable DCB features for Chelsio T4.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

amotin committed:
ntb/plx: Workaround read-only scratchpad registers

On several systems we've noticed that when NTB link goes down, the
Physical Layer User Test Pattern registers we use as additional
scratchpad registers (that is explicitly allowed by the chip specs)
become read-only for about 100us.  I see no explanation for this in
the chip specs, neither why it was not seen before, may be a race.
Since we do need these registers, workaround it by repeating writes
until we succeed or 1ms timeout expire.

ixhamza committed:
NVME Encapsulation support for Trimode HBA

This adds support to emulate NVME functionality for the drives
connected on Trimode HBA. Userspace tools can talk to the NVME
device through passthrough commands, e.g., nvme-cli can identify
controller using `nvme id-ctrl /dev/sdX`.

Commit:
Pull in Debian Linux kernel v6.6.15 config

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

Commit:
Disable kernel modules compression

Debian has defaulted to compressing the modules in XZ format. This
introduces some changes that are not desired. The debug info/symbols
for modules are not stripped anymore. As a result, the size of modules
increases ten folds. Previosuly, the debug symbols for modules and
kernel were packaged separately in a -dbg package. But after this
config update, -dbg package only contains symbols for kernel binary
itself. Kernel modules have in built debug symbols. The size of ISO
image increses 3-4 times because of this. Revert to preveious state
for this config.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

bmeagherix committed:
Change CONFIG_CONSOLE_LOGLEVEL_DEFAULT to 6 for production build (#157)

anodos325 committed:
NAS-127114 / 24.10 /  Add support for alternate data streams (#164)

SMB Protocol Background:
------------------------
Filesystems presented over the SMB protocol may support alternate data
streams ("named streams") within a file or a directory. This support is
designated by the filesystem attribute FILE_NAMED_STREAMS. Named streams
are not identical to extended attributes (EAs), which may also be
supported by the same SMB server.

A named stream is a place within a file in addition to the main stream
(normal file data) where data is stored. Named streams have different
data than the main stream (and than each other) and may be written and
read independently of each other. Named streams for a file are designated
by appending a ":" colon character to the file name followed by the
name of the alternate data stream. Stream names may be no more than 255
characters in length and are subject to the characteristics and
limitations documented in MS-FSCC Section 2.1.5 Pathname and following.

A list of named streams for a file can be gathered by submitting an
SMB2_QUERY_INFO request for FILE_STREAM_INFORMATION. The expected server
response is documented in MS-FSS Section 2.4.43 FileStreamInformation.

Streams are typically smallish in size (less than 200 bytes individually),
and are rarely used apart from MacOS SMB clients.

TrueNAS /ZFS background:
------------------------
Solaris supported a similar feature set through its file-backed xattr
capabilities and APIs. This meant that the kernel SMB server in solaris
was able to seamlessly provide support for named streams. When ZFS was
ported to FreeBSD and Linux the extattr and xattr OS APIs were layered
on top of the ZFS file-backed xattrs. As time progressed and ZFS on
Linux saw more use, it was determined that the performance and lack of
atomicity of operations on file-backed xattrs was insufficient for
some application requirements (this was especially the case for Samba
shares), this eventually led to the ZFS dataset configuration parameter
for SA-backed xattrs on Linux (which is the TrueNAS default). With this
configuration, xattrs up to a certain size are written as SA, and larger
xattrs are written as files. The practical result of this is that
TrueNAS can support extended attributes that are much greater in size
than a traditional Linux file server. Unfortunately, due to inability
to perform partial reads and writes on extended attributes a 2 MiB
upper bound is placed as the maximum size of a single extended attribute
/ named stream in TrueNAS.

Samba background:
-----------------
Samba has the ability to present extended attributes as named streams
to SMB clients. This is achieved by prepending a special prefix to
the extended attribute (to differentiate the streams xattrs from normal
xattrs that are presented as EAs over the SMB protocol). Due to
historical design decisions, the Samba module in charge of translating
xattrs into streams appends an extra NULL byte to the xattr on writes
to the local filesystem and strips it off when converting to a stream
for SMB clients.

Implementation details:
-----------------------
This commit adds support for the Linux kernel SMB2/3 client to enumerate
streams on a remote SMB server by including them in the output of
listxattr with the special Samba prefix. Streams may be written to
the remote SMB server via setxattr and read through getxtattr. The
Samba-specific behavior for appending / removing an extra byte to
the xattr can be disabled by setting /proc/fs/cifs/stream_samba_compat
to 0.

Commit:
pci: Enable overrides for missing ACS capabilities

PCIe ACS (Access Control Services) is the PCIe 2.0+ feature that
allows us to control whether transactions are allowed to be redirected
in various subnodes of a PCIe topology.  For instance, if two
endpoints are below a root port or downsteam switch port, the
downstream port may optionally redirect transactions between the
devices, bypassing upstream devices.  The same can happen internally
on multifunction devices.  The transaction may never be visible to the
upstream devices.

One upstream device that we particularly care about is the IOMMU.  If
a redirection occurs in the topology below the IOMMU, then the IOMMU
cannot provide isolation between devices.  This is why the PCIe spec
encourages topologies to include ACS support.  Without it, we have to
assume peer-to-peer DMA within a hierarchy can bypass IOMMU isolation.

Unfortunately, far too many topologies do not support ACS to make this
a steadfast requirement.  Even the latest chipsets from Intel are only
sporadically supporting ACS.  We have trouble getting interconnect
vendors to include the PCIe spec required PCIe capability, let alone
suggested features.

Therefore, we need to add some flexibility.  The pcie_acs_override=
boot option lets users opt-in specific devices or sets of devices to
assume ACS support.  The "downstream" option assumes full ACS support
on root ports and downstream switch ports.  The "multifunction"
option assumes the subset of ACS features available on multifunction
endpoints and upstream switch ports are supported.  The "id:nnnn:nnnn"
option enables ACS support on devices matching the provided vendor
and device IDs, allowing more strategic ACS overrides.  These options
may be combined in any order.  A maximum of 16 id specific overrides
are available.  It's suggested to use the most limited set of options
necessary to avoid completely disabling ACS across the topology.

Note to hardware vendors, we have facilities to permanently quirk
specific devices which enforce isolation but not provide an ACS
capability.  Please contact me to have your devices added and save
your customers the hassle of this boot option.

Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
Committed-by: Umer Saleem <usaleem@ixsystems.com>

nj-shetty committed:
block: Introduce queue limits and sysfs for copy-offload support

Add device limits as sysfs entries,
	- copy_max_bytes (RW)
	- copy_max_hw_bytes (RO)

Above limits help to split the copy payload in block layer.
copy_max_bytes: maximum total length of copy in single payload.
copy_max_hw_bytes: Reflects the device supported maximum limit.

Signed-off-by: Nitesh Shetty <nj.shetty@samsung.com>
Signed-off-by: Kanchan Joshi <joshi.k@samsung.com>
Signed-off-by: Anuj Gupta <anuj20.g@samsung.com>
Reviewed-by: Hannes Reinecke <hare@suse.de>

nj-shetty committed:
Add infrastructure for copy offload in block and request layer.

We add two new opcode REQ_OP_COPY_DST, REQ_OP_COPY_SRC.
Since copy is a composite operation involving src and dst sectors/lba,
each needs to be represented by a separate bio to make it compatible
with device mapper.
We expect caller to take a plug and send bio with destination information,
followed by bio with source information.
Once the dst bio arrives we form a request and wait for source
bio. Upon arrival of source bio we merge these two bio's and send
corresponding request down to device driver.
Merging non copy offload bio is avoided by checking for copy specific
opcodes in merge function.

Signed-off-by: Nitesh Shetty <nj.shetty@samsung.com>
Signed-off-by: Anuj Gupta <anuj20.g@samsung.com>

nj-shetty committed:
block: add copy offload support

Introduce blkdev_copy_offload to perform copy offload.
Issue REQ_OP_COPY_DST with destination info along with taking a plug.
This flows till request layer and waits for src bio to arrive.
Issue REQ_OP_COPY_SRC with source info and this bio reaches request
layer and merges with dst request.
For any reason, if a request comes to the driver with only one of src/dst
bio, we fail the copy offload.

Larger copy will be divided, based on max_copy_sectors limit.

Reviewed-by: Hannes Reinecke <hare@suse.de>
Signed-off-by: Anuj Gupta <anuj20.g@samsung.com>
Signed-off-by: Nitesh Shetty <nj.shetty@samsung.com>

Commit:
fs/read_write: Enable copy_file_range for block device.

This is a prep patch. Allow copy_file_range to work for block devices.
Relaxing generic_copy_file_checks allows us to reuse the existing infra,
instead of adding a new user interface for block copy offload.
Change generic_copy_file_checks to use ->f_mapping->host for both inode_in
and inode_out. Allow block device in generic_file_rw_checks.

Reviewed-by: Hannes Reinecke <hare@suse.de>
Signed-off-by: Anuj Gupta <anuj20.g@samsung.com>
Signed-off-by: Nitesh Shetty <nj.shetty@samsung.com>

nj-shetty committed:
fs, block: copy_file_range for def_blk_ops for direct block device

For direct block device opened with O_DIRECT, use blkdev_copy_offload to
issue device copy offload, or use splice_copy_file_range in case
device copy offload capability is absent or the device files are not open
with O_DIRECT.

Reviewed-by: Hannes Reinecke <hare@suse.de>
Signed-off-by: Anuj Gupta <anuj20.g@samsung.com>
Signed-off-by: Nitesh Shetty <nj.shetty@samsung.com>

ixhamza committed:
nvme: acquire a new subsystem if no valid controllers assigned

yocalebo committed:
disable multiple device (md) driver

ixhamza committed:
ses: fallback to read SAS address from VPD

The mpt3sas driver does not create a parent end device for PCIe types
where the SAS address is stored, causing the ses driver to not add
PCIe device types connected to a tri-mode HBA. To address this, the
fallback mechanism reads the SAS address from the VPD 0x83 page. This
change is inspired by commit 9927c68.

ixhamza committed:
Add support for two bdevs in blkdev_copy_offload

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

anodos325 committed:
NAS-130052 / 24.10 / Fix regression in stream truncation (#183)

Upstream commit 6ebfede8d57a615dcbdec7e490faed585153f7f1 changed
API for SMB2_set_eof() which introduced regression in truncation
of SMB alternate data streams.

Commit:
Enable CONFIG_WERROR for TrueNAS kernels

Enable CONFIG_ERROR option that treats all warnings as error during
kernel build.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

Commit:
Remove err_dma_mask label from ntb_hw_gen1.c

Label err_dma_mask is not being used and generates a warning at build
time. With CONFIG_WERROR enabled, this warning is treated as error and
breaks the build. This commit removes this label for now.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

anodos325 committed:
Bump up maximum allowed ACL entries (#189)

nfsd supports up to 1024 ACEs. Windows servers support a security
descriptor size of up to 64 KiB, which translates to around 1700
aces, but since we don't locally support that size we'll keep both
at 1024 max.

Commit:
Fix installation of symlink for headers in modules directory

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

ixhamza committed:
scsi: mpi3mr: NVMe encapsulation support for tri-mode HBA

Commit:
Import Debian config for Linux v6.12.5

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

Commit:
Disable kernel modules compression

The config option that disable the module comdule compression is
updated in 6.12.5.

Signed-off-by: Umer Saleem <usaleem@ixsystems.com>

ixhamza committed:
scsi: mpi3mr: restrict scsih IOCTLs for NVMe

bmeagherix committed:
NAS-134656 / 25.10 / Add NVMe host and target auth (#211)

* Add NVMe host and target auth

* Add NVME_TARGET_DEBUGFS

ixhamza committed:
hwmon: (drivetemp) Add SCSI drives support

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
Enable in-tree module signing with ephemeral key

The kernel will auto-generate a keypair at build time if
`certs/signing_key.pem` does not exist.

ixhamza committed:
Enable lockdown in case of Secure Boot

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
ses: Add NVMe support to SCSI Enclosure Services

Add support for NVMe devices in SCSI Enclosure Services. NVMe devices
are supported through both hotplug detection and polling. This allows
NVMe controller devices to appear in enclosure sysfs under their
corresponding slots.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

anodos325 committed:
NAS-132120 / 25.10 / expose ZFS snapdir over NFS (#207) (#208)

This commit adds support for auto-mounting ZFS snapshot directories
as they are accessed over the NFS protocol if the export option
NFSEXP_SNAPDIR has been set.

Signed-off-by: Andrew Walker <awalker@ixsystems.com>

mgrimesix committed:
Wrap is_snapdir in CONFIG_TRUENAS.

ixhamza committed:
ses: fix NULL dereference accessing sysfs during enclosure removal

A race condition exists when ses_intf_remove_enclosure() sets
edev->scratch to NULL while userspace is accessing sysfs attributes.
The sysfs read callbacks access edev->scratch to get ses_device,
causing NULL pointer dereference and kernel panic. Fix this by calling
enclosure_unregister() before clearing edev->scratch. Since
enclosure_unregister() synchronously removes all sysfs entries via
device_unregister(), no further sysfs access is possible when freeing
the ses_device data structures, eliminating the race.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
scsi: mpi3mr: Fix NVMe ioctl handling for non-PCIe devices

The driver incorrectly allows NVMe ioctls on SAS/SATA devices because it
accesses dev_spec.pcie_inf.dev_info without first checking dev_type. For
SAS/SATA devices, the dev_spec union contains sas_sata_inf data, but the
driver reads it as pcie_inf, misinterpreting the data and potentially
matching the NVMe type check, causing NVME_IOCTL_ID to incorrectly
return 1 (SUCCESS) instead of -EINVAL. The only place Jeff is able to
reproduce this is M60 Active controller where sedutil-cli --scan were
showing NVMe identify error for SAS disks on 9600 HBA. Add a dev_type
check to ensure only PCIe devices proceed with NVMe ioctl handling.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
Standardize CONFIG_TRUENAS checks to use #ifdef instead of #if

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
CI: Add libdw-dev to debug kernel dependencies for 6.18

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
CI: Fix disk space issues when building 6.18 kernel

Add more aggressive cleanup of unused system directories to free up
additional disk space. The 6.18 kernel build was requiring more space
due to significantly larger build artifacts compared to 6.12.

Added cleanup of Android SDK, Swift, GHCup directories, Docker images,
and APT cache to provide several additional gigabytes of free space.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
Import Debian config for Linux v6.18-rc7

Import configuration from Debian's
linux-config-6.18_6.18~rc7-1~exp1_amd64.deb package. At the time of
the merge, 6.18-rc7 is the most recent kernel config available in
Debian's package repository
(https://ftp.debian.org/debian/pool/main/l/linux/). Will update to
the final 6.18 release config once available.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
CI: Cancel outdated workflow runs instead of newer ones

Change concurrent_skipping from 'same_content_newer' to 'outdated_runs'.
Previously, when pushing multiple commits in quick succession, the newer
workflow run would be skipped while the older one continued. This is
backwards for development, i.e., we want to test the latest code and
cancel the outdated runs.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
Add TrueNAS Debian build customizations

This commit adds TrueNAS build customization required for building
Debian packages for TrueNAS SCALE kernel.

The original commit ported from v6.6 is
0c5b36a432e1f86ee8dbd9a80aca8dd0919b989e.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
Enable CONFIG_NFSD_LEGACY_CLIENT_TRACKING for 6.18

Debian disabled this in 6.18 as deprecated, breaking API test
test_300_nfs test_state_directory which requires the
/proc/fs/nfsd/nfsv4recoverydir interface. Re-enable for API test
compatibility.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza committed:
Disable CONFIG_TRACEFS_AUTOMOUNT_DEPRECATED to suppress kernel warnings

Kernel 9ba817fb7c6a deprecated automounting tracefs in debugfs, causing
"Automounting of tracing to debugfs is deprecated and will be removed in
2030" warnings whenever /sys/kernel/debug/tracing is accessed. This
warning spams kernel logs during normal operation. Tools should use
/sys/kernel/tracing instead of /sys/kernel/debug/tracing.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

bugclerk commented:
Jira URL: https://ixsystems.atlassian.net/browse/NAS-137494

yocalebo reviewed (approved):
I trust you've done as much testing as humanly possible

mgrimesix reviewed (approved):
I've opened https://ixsystems.atlassian.net/browse/NAS-138841 for fixing the `CONFIG_NFSD_LEGACY_CLIENT_TRACKING=y` config addition.

bmeagherix reviewed (approved):
Ran the sharing protocols tests with the build.  Results [here](http://jenkins.eng.ixsystems.net:8080/job/tests/job/sharing_protocols_tests/1060/).

iSCSI and NVMe-oF look fine, so I will approve.  (Also spun up a VM to ensure HA internode iSCSI disk suppression was functional).

However there seem to be some SMB client regressions.

ixhamza committed:
cifs: Fix TrueNAS ZFS ACL xattr retrieval with kernel 6.18+

Upstream Commit ad9364a6835c changed get_acl() to request only the
security information specified in extra_info parameter, instead of
always including OWNER|GROUP|DACL. The TrueNAS system.nfs4_acl_xdr
handler was passing 0 as extra_info, causing the server to return an
empty Security Descriptor with dacloffset == 0. This triggered the null
ACL generation path, returning a single everyone@ entry (28 bytes)
instead of the actual ACL entries from the server.

Fix by explicitly requesting OWNER_SECINFO | GROUP_SECINFO |
DACL_SECINFO in the XATTR_ZFSACL case, ensuring the full Security
Descriptor is retrieved.

Signed-off-by: Ameer Hamza <ahamza@ixsystems.com>

ixhamza commented:
> However there seem to be some SMB client regressions.

Thanks @bmeagherix for running sharing protocol tests. All the smb client tests were being affected by 6.18 merge. The issue was reliably reproducible on 6.18 client side as we get 28 bytes entry instead of full 48-88 bytes entry depending on SID mapping. After digging deeper, found out that ad9364a6835c caused the regression. The commit changed `SMB2_query_acl()` to not automatically include `OWNER`|`GROUP`|`DACL`, instead it only uses what's passed in `extra_info`. TrueNAS code was passing 0, so server returned empty Security Descriptor.

I have added a fix according to my understanding of the issue: f1a88a6
It explicitly requests `OWNER_SECINFO` | `GROUP_SECINFO` | `DACL_SECINFO` in the `XATTR_ZFSACL` case. Waiting for [Scale Build](http://jenkins.eng.ixsystems.net:8080/job/master/job/custom/1751/) to complete, and will run sharing Sharing Protocol Tests afterwards.

But it needs deeper eyes by @anodos325 to make sure this is the right approach for handling the upstream change.

ixhamza commented:
Triggered the following test runs. Will merge once the test runs complete with expected results:
- [API Tests](http://jenkins.eng.ixsystems.net:8080/job/tests/job/api_tests/6890/console)
- [Sharing Protocol Tests](http://jenkins.eng.ixsystems.net:8080/job/tests/job/sharing_protocols_tests/1069/console)
- [Unit Tests](http://jenkins.eng.ixsystems.net:8080/job/tests/job/unit_tests/2720/console)
- [Directory Services Tests](http://jenkins.eng.ixsystems.net:8080/job/tests/job/directoryservices_tests/1459/console)
- [Stig Tests](http://jenkins.eng.ixsystems.net:8080/job/tests/job/stig_tests/1173/console)
- [Cloud Tests](http://jenkins.eng.ixsystems.net:8080/job/tests/job/cloud_tests/663/console)

bugclerk commented:
Not updating JIRA ticket https://ixsystems.atlassian.net/browse/NAS-137494 target versions as no JIRA version corresponds to this PR

bugclerk commented:
This PR has been merged and conversations have been locked.
If you would like to discuss more about this issue please use our forums or raise a Jira ticket.
